hack/tools/metal.py

```diff
@@ -0,0 +1,171 @@
+      project = self.create_new_project_if_not_exist(prj_name_or_id)
+      self.project = project
+      key = self.create_new_key(project.id, key_name)
+      self.key = key
+      if userdata == None:
+         userdata = self.create_user_data()
+      device = self.create_device(project.id, dev_name, key.id, userdata)
+      self.device = device
+      self.dev_id = device.id
+      self.wait_until_device_ready(device)
+
+      return self.ip
+
+   def create_new_project_if_not_exist(self, name_or_id):
+      project = None
+      try:
+         project = self.packet_manager.get_project(name_or_id)
+         self.logger.info(f"using found project {name_or_id}")
+         return project
+      except Exception:
+         pass
+
+      project = self.packet_manager.create_organization_project(
+          org_id=self.org_id,
+          name=name_or_id
+      )
+      self.logger.info(f"created project {name_or_id}")
+
+      return project
+
+   def create_new_key(self, project_id, name):
+      key = RSA.generate(2048)
+      with open(self.private_key_path, 'wb') as priv_file:
+          os.chmod(self.private_key_path, 0o600)
+          priv_file.write(key.exportKey('PEM'))
+
+      pubkey = key.publickey().exportKey('OpenSSH')
+      with open(self.public_key_path, 'wb') as pub_file:
+          pub_file.write(pubkey)
+
+      key = self.packet_manager.create_project_ssh_key(project_id, name, pubkey.decode("utf-8"))
+      self.logger.info(f"created key {key.label}")
+
+      return key
+
+   def create_user_data(self):
+      files = ["hack/scripts/bootstrap.sh", "hack/scripts/devpool.sh", "test/e2e/infra/userdata.sh"]
+      userdata = ""
+      for file in files:
+         with open(self.base+"/"+file) as f:
+             userdata += f.read()
+
+      return userdata
+
+   def create_device(self, project_id, name, key_id, userdata):
+      device = self.packet_manager.create_device(project_id=project_id,
+                                     hostname=name,
+                                     plan='c1.small.x86', metro='sv',
+                                     operating_system='ubuntu_18_04',
+                                     facility="ewr1", billing_cycle='hourly',
+                                     project_ssh_keys=[key_id],
+                                     userdata=userdata)
+      self.logger.info(f"created device {device.hostname}")
+
+      return device
+
+   # TODO timeouts
+   def wait_until_device_ready(self, device):
+      while True:
+         d = self.packet_manager.get_device(device_id=device.id)
+         self.logger.info(f"checking state of device {d.hostname} ...")
+         if d.state == "active":
+             self.logger.info(f"{d.hostname} running")
+             break
+         self.logger.info("not ready yet, sleeping for 10s")
+         time.sleep(10)
+
+      ips = device.ips()
+      self.ip = ips[0].address
+
+      while True:
+         shell = self.new_shell(self.ip)
```

Comment:
lol that was the first thing i tried, and it was NOT happy 😂  i cant remember the error now, let me move it and make it break again

Reply:
oh i think i remember. because of the whole "using the `with` keyword as a `defer` thing, the connection was closed after just one turn through the loop.
i can switch it to not defer closing and use the same connection each time, but then i have to manually close it, and i am not sure if there is a nice way to do that in a defer-type way so that it definitely closes